blocks/bricks/conv.py

```diff
@@ -0,0 +1,163 @@
+        add_role(W, FILTERS)
+        self.params.append(W)
+        self.add_auxiliary_variable(W.norm(2), name='W_norm')
+
+    def _initialize(self):
+        W, = self.params
+        self.weights_init.initialize(W, self.rng)
+
+    @application(inputs=['input_'], outputs=['output'])
+    def apply(self, input_):
+        """Perform the convolution.
+
+        Parameters
+        ----------
+        input_ : :class:`~tensor.TensorVariable`
+            A 4D tensor with the axes representing batch size, number of
+            channels, image height, and image width.
+
+        Returns
+        -------
+        output : :class:`~tensor.TensorVariable`
+            A 4D tensor of filtered images (feature maps) with dimensions
+            representing batch size, number of filters, feature map height,
+            and feature map width.
+
+            The height and width of the feature map depend on the border
+            mode. For 'valid' it is ``image_size - filter_size + 1`` while
+            for 'full' it is ``image_shape + filter_size - 1``.
+
+        """
+        W, = self.params
+        output = conv2d(
+            input_, W, subsample=self.step, border_mode=self.border_mode,
+            filter_shape=(self.num_filters,
+                          self.num_channels) + self.filter_size)
+        return output
+
+
+class MaxPooling(Initializable, Feedforward):
+    """Max pooling layer.
+
+    Parameters
+    ----------
+    pooling_size : tuple
+        The height and width of the pooling region i.e. this is the factor
+        by which your input's last two dimensions will be downscaled.
+    step : tuple, optional
+        The vertical and horizontal shift (stride) between pooling regions.
+        By default this is equal to `pooling_size`. Setting this to a lower
+        number results in overlapping pooling regions.
+
+    """
+    @lazy
+    def __init__(self, pooling_size, step=None, **kwargs):
+        super(MaxPooling, self).__init__(**kwargs)
+        self.pooling_size = pooling_size
+        self.step = step
+
+    @application(inputs=['input_'], outputs=['output'])
+    def apply(self, input_):
+        """Apply the pooling (subsampling) transformation.
+
+        Parameters
+        ----------
+        input_ : :class:`~tensor.TensorVariable`
+            An tensor with dimension greater or equal to 2. The last two
+            dimensions will be downsampled. For example, with images this
+            means that the last two dimensions should represent the height
+            and width of your image.
+
+        Returns
+        -------
+        output : :class:`~tensor.TensorVariable`
+            A tensor with the same number of dimensions as `input_`, but
+            with the last two dimensions downsampled.
+
+        """
+        output = max_pool_2d(input_, self.pooling_size, st=self.step)
+        return output
+
+
+class ConvolutionalLayer(Sequence, Initializable):
```

Comment:
I don't think there's an easy way besides just doing all the messy calculations that take into account all the possible settings (the border mode, the pool stride, the filter stride, etc.)

But actually the main reason I didn't implement this is because it's not clear to me what the "dimension" of a 2D input or output is. I guess there are basically two options: `height * width` or `(height, width)`. Then the question is whether they should share the interface with `FeedForward` (which so far has always assumed that `dim` is an integer). Curious to hear what @dwf and @rizar have to say about that.